Re-merge #63668: Improve @objcImplementation member checking

beccadax opened the pull request:
This PR undoes the revert of #63668 and adds additional changes to fix a bug blocking integration and improve tests. The original:

> This PR introduces a new, more robust checker for the members of `@_objcImplementation` extensions and makes related Sema-level improvements. With it, swiftc can now:
> 
> * Infer `@objc` on member implementations.
> * Emit diagnostics which describe currently-proposed semantics (e.g. using `fileprivate` and `private` to create `@objc` helper methods).
> * Globally handle name matching (for instance, if there are two ObjC methods with the same Swift name, adding `@objc(<explicit selector>)` to one of the Swift implementations will make both of them match the right method).
> * Diagnose and offer fix-its for mismatched names in some situations.
> * Allow explicit `deinit`s in `@_objcImplementation extension`s.
> 
> This PR is also capable of detecting ambiguous matches, but these situations are not being diagnosed yet because I haven't figured out how to avoid burying users in a blizzard of diagnostics.
> 
> Fixes rdar://103150189.

This PR additionally:

* Excludes `@objcImplementation` member implementations from being included in module interfaces, fixing the issue that originally forced the revert.
* Re-enables the IRGen/objc_implementation.swift test on macOS, where it's known to work correctly.
* Tweaks the test modifications in IRGen/objc_implementation.swift relating to `deinit` support.
* **NEW:** Fixes an `@objcImplementation`-specific bug in the logic which decides whether to print the `override` keyword in a module interface.

beccadax committed:
[NFC] Expose extension category name accessor

beccadax committed:
Correct and improve objcImpl member matching

• `@objc` is now inferred on non-`final` members of @objcImplementation extensions
• Diagnostics now suggest adding `private` to ObjC helper members, not `@objc`, in line with currently proposed behavior
• Better diagnostic for members implemented in the wrong extension

Part of rdar://103150189.

beccadax committed:
Diagnose objcImpl members with wrong name

beccadax committed:
Refactor and expand @objcImpl checking

Create a checker for @_objcImplementation member implementations that considers all of a class’s interface and implementation decls at once. This allows us to handle several things better:

• Unimplemented requirements are now diagnosed

• Header members that can match several implementations, or implementations that could match several header members, are now diagnosed

• Tailored diagnostic when the implementation's Swift name matches the header's selector instead of its Swift name

• Recommends inserting `@objc(<selector>)` when a Swift name matches but the implicit ObjC name doesn't

• An `@objc(<selector>)` on one implementation can eliminate its requirement from being considered for other implementations, resolving ambiguities

This does unfortunately regress the diagnostics when a requirement is implemented in the wrong extension. Some sort of whole-module checking would be needed to address this problem.

beccadax committed:
Better support destructors in objcImpl

beccadax committed:
Omit @objcImpl from module interfaces

Module interfaces should not include the @objcImplementation attribute, member implementations that are redundant with the ObjC header, or anything that would be invalid in an ordinary extension (e.g. overridden initializers, stored Swift-only properties).

beccadax committed:
Partly re-enable IRGen/objc_implementation.swift

Partly fixes rdar://101420862, and catches up the test with subsequent changes.

tshortli reviewed (commented):
I took a look at the module interface bits and they look good to me! Just one question and a minor request.

beccadax commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -137,6 +137,11 @@ static bool contributesToParentTypeStorage(const AbstractStorageDecl *ASD) {
 }
 
+static bool isInObjCImpl(const ValueDecl *VD) {
+  auto *ED = dyn_cast<ExtensionDecl>(VD->getDeclContext());
```
I believe it only needs to apply to direct members—if you somehow declared, say, a nested type inside an `@_objcImplementation` extension (which would be pretty weird in the first place), you wouldn't want to apply special logic to its members. Recall that accessors are represented in the AST as peers of their storage, not members of their storage, and that one extension cannot be nested inside another. If you think there's a case I'm not thinking of, let me know!

tshortli replied:
Ok, makes sense to me; the members of the nested type wouldn't be fulfilling member requirements declared by the header because you can't nest declarations like that in Obj-C.

beccadax commented on test/ModuleInterface/objc_implementation.swift:
```diff
@@ -0,0 +1,84 @@
+// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -emit-module -I %S/Inputs/objc_implementation -F %clang-importer-sdk-path/frameworks %s -import-underlying-module -swift-version 5 -enable-library-evolution -emit-module-interface-path %t.swiftinterface
+// RUN: %FileCheck --input-file %t.swiftinterface %s
```
Good catch—this turned up an issue with a missing `override` keyword that I'll need to fix.

beccadax replied:
The bug I discovered has now been fixed and the test is using the standard helpers to emit and verify the interface.

beccadax committed:
Fix @objcImpl subclass init visibility computation

When writing an @objc subclass of an @objcImplementation class, implicit initializers in the subclass were treated as overriding the *implementation decl*, not the *interface decl*, of the initializer in the superclass. This caused Swift to incorrectly compute the visibility of the superclass initializer and omit an `override` keyword from the module interface when one would definitely be necessary.

Correct this oversight by looking up the interface decl matching the superclass implementation in `collectNonOveriddenSuperclassInits()`.

beccadax committed:
[NFC] Verify ModuleInterface/objc_implementation

Switch the ModuleInterface/objc_implementation.swift test to using standard substitutions to properly emit its module interface, and make sure it also verifies.

beccadax commented:
@swift-ci test

beccadax commented:
Tightened up some assertions in a test to avoid false negatives.

@swift-ci please test

beccadax commented:
@swift-ci please smoke test and merge

beccadax commented:
Rebased to fix a merge conflict.

beccadax commented:
@swift-ci smoke test and merge

beccadax commented:
@swift-ci smoke test and merge

nkcsgexi commented:
Can we also merge this to 5.9 branch?
